Neo4j.Driver/Neo4j.Driver/Internal/Extensions/CollectionExtensions.cs

```diff
@@ -86,8 +88,46 @@ public static string ValueToString(this object o)
 
         public static IDictionary<string, object> ToDictionary(this object o)
         {
-            return o.GetType().GetRuntimeProperties()
-                .ToDictionary(prop => prop.Name, prop => prop.GetValue(o, null));
+            if (o == null)
+            {
+                return null;
+            }
+
+            return FillDictionary(o, new Dictionary<string, object>());
+        }
+
+        private static IDictionary<string, object> FillDictionary(object o, IDictionary<string, object> dict)
+        {
+            foreach (var propInfo in o.GetType().GetRuntimeProperties())
+            {
+                var name = propInfo.Name;
+                var value = propInfo.GetValue(o);
+
+                if (value != null)
+                {
+                    var convertible = value as IConvertible;
+                    if (convertible != null)
```

Comment:
Introduced recursive conversion of objects in the object hierarchy - supporting arrays, lists and dictionaries. Also added a number of integration tests that test against possible use cases documented [here](http://neo4j.com/docs/developer-manual/current/cypher/syntax/parameters/).